src/appleseed/renderer/modeling/postprocessingstage/bloompostprocessingstage.cpp

```diff
@@ -0,0 +1,425 @@
+                    max_level_props.m_pixel_format);
+
+                blur_pyramid_down.push_back(Image(level_props));
+                blur_pyramid_up.push_back(Image(level_props));
+
+                // Halve dimensions for the next buffer level.
+                assert(level_width >= 4 && level_height >= 4);
+                level_width /= 2;
+                level_height /= 2;
+            }
+        }
+
+        void execute(Frame& frame, const std::size_t thread_count) const override
+        {
+            if (m_iterations == 0 || (m_intensity == 0.0f && !m_debug_blur))
+                return;
+
+            Image& image = frame.image();
+            const CanvasProperties& props = image.properties();
+
+            // Determine the dimensions of the largest temporary buffer used for blurring.
+            const std::size_t max_pyramid_width = props.m_canvas_width / 2;
+            const std::size_t max_pyramid_height = props.m_canvas_height / 2;
+            const std::size_t min_pyramid_dimension = std::min(max_pyramid_width, max_pyramid_height);
+
+            if (min_pyramid_dimension < 4)
+                return;
+
+            // Copy the pixel format, but ignore the alpha channel.
+            assert(props.m_channel_count == 4);
+            const std::size_t max_tile_size = 32;
+            const CanvasProperties blur_props(
+                max_pyramid_width,
+                max_pyramid_height,
+                std::min(max_pyramid_width, max_tile_size),
+                std::min(max_pyramid_height, max_tile_size),
+                3,
+                props.m_pixel_format);
+
+            // Compute how many downsampling iterations we can do before a buffer has a side smaller than 2 pixels.
+            const float max_iterations = std::log2(min_pyramid_dimension / 2.0f);
+            const std::size_t iterations = clamp<std::size_t>(m_iterations, 1, static_cast<std::size_t>(max_iterations));
+
+            if (iterations < m_iterations)
+            {
+                RENDERER_LOG_INFO(
+                    "post-processing stage \"%s\":\n"
+                    "  the frame is too small for " FMT_SIZE_T " iterations\n"
+                    "  using " FMT_SIZE_T " iterations instead",
+                    get_path().c_str(),
+                    m_iterations,
+                    iterations);
+            }
+
+            //
+            // Prefilter pass.
+            //
+
+            // Copy the original image but ignore its alpha channel.
+            const std::size_t shuffle_table[4] = { 0, 1, 2, Pixel::SkipChannel };
+            Image prefiltered_image(image, props.m_pixel_format, shuffle_table);
+
+            // Filter out dark pixels.
+            const BrightPassApplier bright_pass(m_threshold, m_soft_threshold);
+            bright_pass.apply_on_tiles(prefiltered_image, thread_count);
+
+            //
+            // Create and initialize the blur buffer pyramids.
```

Comment:
Can you explain why we need such buffer ?

Reply:
When we blur an image its intensity fades. So, by adding the intermediate blurring steps we can avoid this, e.g.:
![Jfw0rZ5bno](https://user-images.githubusercontent.com/13294013/87185437-25afa980-c2c0-11ea-8612-bf986fcf2aa2.gif)